### PR TITLE
Upgrade pyo3 to 0.23, sync changelogs

### LINKED ISSUE #412: Update pyo3 to version 0.23
Among other improvements, this includes the ability to support freethreaded python in 3.13

### PR DESCRIPTION
**Pull request by mpiannucci**
There are some deprecations that directly affect our AsyncGenerator implementation, so we have to figure that out (EDIT: See #416)

Closes #412

### CODE REVIEW

File: icechunk-python/python/icechunk/__init__.py [hunk 1]
```diff
@@ -22,13 +22,13 @@
...
 
 __all__ = [
-    "__version__",
     "IcechunkStore",
     "S3Credentials",
     "SnapshotMetadata",
     "StorageConfig",
     "StoreConfig",
     "VirtualRefConfig",
+    "__version__",
```
Comment (mpiannucci) [inline icechunk-python/python/icechunk/__init__.py:31]: @paraseba This was the fix you needed. IDK why teh fix wasnt automatically applied or detected locally

**Comment by mpiannucci** [comment]
JK have to figure out deprecations still

File: icechunk-python/src/lib.rs [hunk 2]
```diff
@@ -890,6 +857,7 @@ impl PyIcechunkStore {
     }
 
     fn list(&self) -> PyIcechunkStoreResult<PyAsyncGenerator> {
+        #[allow(deprecated)]
```
Comment (mpiannucci) [inline icechunk-python/src/lib.rs:860]: This is a major change to the way we do streams, punting for now and logging issue

Comment (mpiannucci) [reply icechunk-python/src/lib.rs:860]: https://github.com/earth-mover/icechunk/issues/416

File: deny.toml [hunk 3]
```diff
@@ -22,6 +22,7 @@ allow = [
   "OpenSSL",
   "Unicode-DFS-2016",
   "CC0-1.0",
+  "Unicode-3.0",
```
Comment (mpiannucci) [inline deny.toml:25]: `main` is currently not passing CI because of this.

File: Changelog.python.md [hunk 4]
```diff
@@ -1,5 +1,21 @@
 # Changelog
 
+## Python Icechunk Library 0.1.0a5
+
+### Features
+
+- Sync with zarr 2.0b2. The biuggest change is the `mode` param on `IcechunkStore` methods has been simplified to `read_only`.
```
Comment (dcherian) [inline Changelog.python.md:7]:
```suggestion
- Sync with zarr 3.0b2. The biggest change is the `mode` param on `IcechunkStore` methods has been simplified to `read_only`.
```

File: Changelog.python.md [hunk 5]
```diff
@@ -1,5 +1,21 @@
...
+- Changed `IcechunkStore::distributed_commit` to `IcechunkStore::merge`, which now does *not* commit, but attempts to merge the changes from another store back into the current store.
+- Added a new `icechunk.dask.store_dask` method to write a dask array to an icechunk store. This is required for safely writing dask arrays to an icechunk store.
```
Comment (dcherian) [inline Changelog.python.md:9]:
```suggestion
```
I didn't intend this to be public.

File: Changelog.python.md [hunk 6]
```diff
@@ -1,5 +1,21 @@
...
+- Added a new `icechunk.xarray.to_icechunk` method to convert an xarray dataset to an icechunk store. This and `icehunk.dask.store_dask` are required for safely writing xarray datasets with distributed dask arrays to an icechunk store.
```
Comment (dcherian) [inline Changelog.python.md:10]:
```suggestion
- Added a new `icechunk.xarray.to_icechunk` method to write an xarray dataset to an icechunk store. This is *required* for safely writing xarray datasets with dask arrays to an icechunk store in a distributed or multi-processing context.
```

File: Changelog.python.md [hunk 7]
```diff
@@ -1,5 +1,20 @@
 # Changelog
 
+## Python Icechunk Library 0.1.0a5
+
+### Features
+
+- Sync with zarr 3.0b2. The biggest change is the `mode` param on `IcechunkStore` methods has been simplified to `read_only`.
+- Changed `IcechunkStore::distributed_commit` to `IcechunkStore::merge`, which now does *not* commit, but attempts to merge the changes from another store back into the current store.
+- Added a new `icechunk.xarray.to_icechunk` method to write an xarray dataset to an icechunk store. This is *required* for safely writing xarray datasets with dask arrays to an icechunk store in a distributed or multi-processing context.
```
Comment (dcherian) [inline Changelog.python.md:9]:
```suggestion
- Added a new `icechunk.dask.store_dask` method to write a dask array to an icechunk store. This is required for safely writing dask arrays to an icechunk store.
- Added a new `icechunk.xarray.to_icechunk` method to write an xarray dataset to an icechunk store. This is *required* for safely writing xarray datasets with dask arrays to an icechunk store in a distributed or multi-processing context.
```

My bad, i added it to the docs! I'll send in a PR with a docstring.

**Review by dcherian — APPROVED** [review]

**Pull request merged**